### DISCUSSION TITLE
Problems with detected langauge

### DISCUSSION THREAD

**Discussion by neclimdul** [vimeo/psalm#7845] [Q&A] [answered]
Not sure how to handle this but I've got a library I'm trying to support so I've got a matrix of PHP versions that I test it against. Everything from PHP 7.3 -> 8.1 currently.  

When I run psalm it detects the language as the minimum language I specify in composer.json and until recently that's worked great. Unfortunately some libraries it depend on have started updating to use 8.X specific features. This doesn't break the library because the interfaces are the same and they're tagged to the correct supported versions so the 7.x matrix gets the older version of the library and the 8.x matrix gets the newer version like one would expect.

The problem comes from the fact that psalm detects the version as the _minimum_ version and, despite being told to ignore the vendor directory, scans the used files and triggers parse errors on new features like match in the vendor files.

So I guess that gets to my question which is twofold.
1) Can I tell psalm to use the current version of PHP instead of detecting it from composer.json?
2) Is there a better way to handle this?

**Comment by orklah** [comment, marked as answer]
You absolutely can tell Psalm what version it must use for analysis: https://psalm.dev/docs/running_psalm/configuration/#phpversion

Psalm is not currently completely aware of what is vendor code and what is project code in some particular (but common cases), it would be a pretty big change to make it analyse vendor code in the most recent php version in order to make sure it understand everything. I guess not impossible, but pretty far above what I'm comfortable with right now...

**Reply by neclimdul** [reply]
I guess that's one way to think about it. Its kinda a sticky problem because on the one hand I do want psalm to warn me locally if I'm writing code that will break a supported PHP version. But also in CI I want to be able to test my code against the latest versions of PHP with what ever libraries composer resolves to warn me of possible breaking changes.

This doesn't look pretty but I guess per your suggestion I could do this in my CI?
```
./vendor/bin/psalm --php-version=$(php -r 'echo PHP_VERSION;')
```

**Reply by orklah** [reply]
Well, if you're gonna be stuck with crash when analyzing in lower versions, you may as well only check with a more recent version. IMHO, if you analyze on the lower version where Psalm doesn't crash, it's enough (no need to run psalm on each version)

**Reply by neclimdul** [reply]
that's a fair assessment as well. thanks for the pointers.
